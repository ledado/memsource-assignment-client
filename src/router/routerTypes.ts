export const routerTypes = {
  MAIN: 'main',
  PROJECT_NEW: 'project-new',
  PROJECT_UPDATE: 'project-update',
};
