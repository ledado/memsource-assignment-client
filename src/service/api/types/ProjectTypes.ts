export interface ProjectListResponse {
  _embedded: {
    projects: ProjectItem[];
  };
  _links: any;
}

export interface ProjectItem {
  id: number;
  name: string;
  sourceLanguage: string;
  status: ProjectStatusEnum;
  targetLanguages: string[];
  dateCreated: string;
  dateUpdated: string;
  dateDue: string;
}

export enum ProjectStatusEnum {
  NEW = 'NEW',
  COMPLETED = 'COMPLETED',
  DELIVERED = 'DELIVERED',
}

export enum ProjectFormTypeEnum {
  CREATE = 'CREATE',
  UPDATE = 'UPDATE',
}
