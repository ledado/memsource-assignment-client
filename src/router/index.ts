import Vue from 'vue';
import VueRouter, {RouteConfig} from 'vue-router';
import {errorBus} from '@/service/events/ErrorBus';
import {layoutTypes} from '@/layouts/layoutTypes';
import {routerTypes} from '@/router/routerTypes';

Vue.use(VueRouter);

const routes: Array<RouteConfig> = [
  {
    path: '/',
    name: routerTypes.MAIN,
    component: () => import('../views/Main.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/new',
    name: routerTypes.PROJECT_NEW,
    component: () => import('../views/Project/MainProjectCreate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/:id',
    name: routerTypes.PROJECT_UPDATE,
    component: () => import('../views/Project/MainProjectUpdate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
    props: true,
  },
  {
    path: '*',
    component: () => import('../views/Error/Error404.vue'),
  },
];

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});

// error state clean up
router.afterEach((to, from) => {
  if (from.name && errorBus.state) {
    errorBus.errorCode = undefined;
    errorBus.state = false;
  }
});

export default router;
