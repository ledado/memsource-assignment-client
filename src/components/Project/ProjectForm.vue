<template>
  <v-row no-gutters>
    <v-col cols="12">
      <v-form ref="form" v-model="valid" lazy-validation>
        <v-text-field
          v-model="value.name"
          :rules="nameRules"
          :label="$t('project.form.name.label')"
          outlined
          required
        ></v-text-field>

        <v-select
          v-model="value.sourceLanguage"
          :items="languageItems"
          :rules="sourceLanguageRules"
          :label="$t('project.form.source_language.label')"
          required
          outlined
        ></v-select>

        <v-select
          v-model="value.targetLanguages"
          :items="languageItems"
          :rules="targetLanguageRules"
          :label="$t('project.form.target_languages.label')"
          required
          outlined
          multiple
        ></v-select>

        <v-btn :disabled="!valid || loading" color="success" block @click="submitProject">
          {{ $t(`project.form.submit.${type}`) }}
        </v-btn>
      </v-form>
    </v-col>
  </v-row>
</template>

<script lang="ts">
import {Component, Prop, Vue} from 'vue-property-decorator';
import {ProjectFormTypeEnum, ProjectItem} from '@/service/api/types/ProjectTypes';

@Component
export default class ProjectForm extends Vue {
  @Prop() value!: ProjectItem;
  @Prop() type!: ProjectFormTypeEnum;
  @Prop() loading!: boolean;

  $refs!: {
    form: any;
  };

  valid = true;

  nameRules: any = [];
  sourceLanguageRules: any = [];
  targetLanguageRules: any = [];

  languageItems = ['en', 'cs', 'ru', 'de', 'fi', 'ko', 'hu', 'ja'];

  created() {
    this.nameRules = [
      (v: any) => !!v || this.$t('project.form.name.required').toString(),
      (v: any) => (v && v.length <= 20) || this.$t('project.form.name.max_length').toString(),
    ];
    this.sourceLanguageRules = [(v: any) => !!v || this.$t('project.form.source_language.required').toString()];
    this.targetLanguageRules = [
      (v: any) => v.length > 0 || this.$t('project.form.target_languages.required').toString(),
    ];
  }

  submitProject() {
    if (this.$refs.form.validate()) {
      this.$emit('submit');
    }
  }

  resetForm() {
    this.$refs.form.reset();
  }
}
</script>
<style lang="scss" scoped></style>
