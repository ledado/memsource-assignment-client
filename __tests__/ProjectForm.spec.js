import {shallowMount} from '@vue/test-utils';
import ProjectForm from '@/components/Project/ProjectForm.vue';
import {routerTypes} from '@/router/routerTypes';
import {ProjectFormTypeEnum, ProjectStatusEnum} from '@/service/api/types/ProjectTypes';

describe('ProjectForm.vue', () => {
  it('render component', () => {
    const wrapper = shallowMount(ProjectForm, {
      propsData: {
        value: {
          id: 0,
          name: 'test',
          sourceLanguage: 'en',
          status: ProjectStatusEnum.NEW,
          targetLanguages: ['cs'],
          dateCreated: '2022-02-11T07:16:04.299Z',
          dateUpdated: '2022-02-11T07:16:04.299Z',
          dateDue: '2022-02-11T07:16:04.299Z',
        },
        type: ProjectFormTypeEnum.CREATE,
        loading: false,
      },
      stubs: ['router-link'],
      mocks: {
        $t: (msg) => msg,
      },
      mixins: [
        {
          data: () => {
            return {
              routerTypes,
            };
          },
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
