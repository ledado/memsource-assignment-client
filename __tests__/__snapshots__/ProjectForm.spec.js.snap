// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectForm.vue render component 1`] = `
"<v-row-stub tag=\\"div\\" nogutters=\\"true\\">
  <v-col-stub cols=\\"12\\" tag=\\"div\\">
    <v-form-stub lazyvalidation=\\"true\\" value=\\"true\\">
      <v-text-field-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"[Function],[Function]\\" successmessages=\\"\\" value=\\"test\\" backgroundcolor=\\"\\" label=\\"project.form.name.label\\" loaderheight=\\"2\\" clearicon=\\"$clear\\" outlined=\\"true\\" type=\\"text\\" required=\\"\\"></v-text-field-stub>
      <v-select-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"[Function]\\" successmessages=\\"\\" value=\\"en\\" appendicon=\\"$dropdown\\" backgroundcolor=\\"\\" label=\\"project.form.source_language.label\\" loaderheight=\\"2\\" clearicon=\\"$clear\\" outlined=\\"true\\" type=\\"text\\" valuecomparator=\\"[Function]\\" nodatatext=\\"$vuetify.noDataText\\" items=\\"en,cs,ru,de,fi,ko,hu,ja\\" itemcolor=\\"primary\\" itemdisabled=\\"disabled\\" itemtext=\\"text\\" itemvalue=\\"value\\" menuprops=\\"[object Object]\\" required=\\"\\"></v-select-stub>
      <v-select-stub errorcount=\\"1\\" errormessages=\\"\\" messages=\\"\\" rules=\\"[Function]\\" successmessages=\\"\\" value=\\"cs\\" appendicon=\\"$dropdown\\" backgroundcolor=\\"\\" label=\\"project.form.target_languages.label\\" loaderheight=\\"2\\" clearicon=\\"$clear\\" outlined=\\"true\\" type=\\"text\\" valuecomparator=\\"[Function]\\" nodatatext=\\"$vuetify.noDataText\\" items=\\"en,cs,ru,de,fi,ko,hu,ja\\" itemcolor=\\"primary\\" itemdisabled=\\"disabled\\" itemtext=\\"text\\" itemvalue=\\"value\\" menuprops=\\"[object Object]\\" multiple=\\"true\\" required=\\"\\"></v-select-stub>
      <v-btn-stub color=\\"success\\" tag=\\"button\\" activeclass=\\"\\" block=\\"true\\" type=\\"button\\">
        project.form.submit.CREATE
      </v-btn-stub>
    </v-form-stub>
  </v-col-stub>
</v-row-stub>"
`;
